probe/endpoint/resolver.go

```diff
@@ -0,0 +1,69 @@
+package endpoint
+
+import (
+	"net"
+	"time"
+
+	"github.com/bluele/gcache"
+)
+
+const (
+	rAddrCacheLen        = 500 // Default cache length
+	rAddrBacklog         = 1000
+	rAddrCacheExpiration = 30 * time.Minute
+)
+
+// ReverseResolver is a caching, reverse resolver
+type ReverseResolver struct {
+	addresses chan string
+	cache     gcache.Cache
+}
+
+// NewReverseResolver starts a new reverse resolver that
+// performs reverse resolutions and caches the result.
+func NewReverseResolver(cacheLen int) *ReverseResolver {
+	r := ReverseResolver{
+		addresses: make(chan string, rAddrBacklog),
+		cache:     gcache.New(cacheLen).LRU().Expiration(rAddrCacheExpiration).Build(),
+	}
+
+	go r.run()
+	return &r
+}
+
+// Get the reverse resolution for an IP address if already in the cache, gcache.NotFoundKeyError otherwise
+// Note: it returns one of the possible names that can be obtained for that IP
+func (r *ReverseResolver) Get(address string) (string, error) {
+	val, err := r.cache.Get(address)
+	if err == nil {
+		return val.(string), nil
+	}
+
+	if err == gcache.NotFoundKeyError {
+		// we trigger a asynchronous reverse resolution when not cached
+		select {
+		case r.addresses <- address:
+		default:
+		}
+	}
+	return "", err
+}
+
+func (r *ReverseResolver) run() {
+	throttle := time.Tick(time.Second / 10)
+	for address := range r.addresses {
+		<-throttle // rate limit our DNS resolutions
+		names, err := net.LookupAddr(address)
```

Comment:
Thats not what I'm referring to; I'm refer to the fact we enqueue a request on the channel after checking the cache, then some time later do the request.  In the time between the two we might have already put the entry in the cache (from a previous request).

Therefore we need to check the cache a second time before actually doing the request.